packages/base/source/adt/union/union.js

```diff
@@ -82,8 +108,11 @@ instead to check if a value belongs to the ADT variant.`);
        *   where 'b = 'a[`@@folktale:adt:tag]
        */
       matchWith(pattern) {
-        return pattern[name](this);
-      } 
+        assertObject('Union.mapObject#matchWith', pattern);
+        if (name in pattern) { return pattern[name](this); }
+        else if (ANY in pattern) { return pattern[ANY](this) }
```

Comment:
In the default branch of this pattern match, we can't always be sure about which variant we're going to get passed in as `this`. For example:

```js
const { union, any } = require('folktale/adt/union');

const List = union('List', {
  Cons: (value, rest) => ({ value, rest }),
  Empty: () => {}
});

const addOneToHead = xs => xs.matchWith({
  [any]: ({ value }) => value + 1
});

const listWithHead = Cons(4, Empty());
addOneToHead(listWithHead);
//=> 5

const emptyList = Empty();
addOneToHead(emptyList);
//=> TypeError: Cannot match against 'undefined' or 'null'.
```

The error here is in the `({ value }) => value + 1` function that tries to destructure a `value` property from `undefined`. Usually, the "default" branch of a pattern match does not depend on the variant that happens to have reached that branch. What do you think about a change like this?

```patch
diff --git a/packages/base/source/adt/union/union.js b/packages/base/source/adt/union/union.js
index 91e98ab..80a2f5c 100644
--- a/packages/base/source/adt/union/union.js
+++ b/packages/base/source/adt/union/union.js
@@ -110,7 +110,7 @@ instead to check if a value belongs to the ADT variant.`);
       matchWith(pattern) {
         assertObject('Union.mapObject#matchWith', pattern);
         if (name in pattern) { return pattern[name](this); }
-        else if (ANY in pattern) { return pattern[ANY](this) }
+        else if (ANY in pattern) { return pattern[ANY](); }
         else { throw new Error(getMatchWithErrorMessage(pattern, name)) }
       }
     });
```